text/000-wasm-bindgen-trait-method-inheritance.md

```diff
@@ -0,0 +1,384 @@
+though the implementation is very different.
+
+# Motivation
+[motivation]: #motivation
+
+Class-based inheritance is very common in JavaScript, especially with the
+[WebIDL bindings](https://heycam.github.io/webidl/#dfn-inherit).
+
+Given how prelevant this pattern is in JavaScript, it would be very useful to
+have a convenient and ergonomic way to make use of this inheritance while
+maintaining optimal code reuse in Rust.
+
+It is currently possible to use method inheritance, but it is very awkward, because
+it requires type annotations with `.into()`:
+
+```rust
+let x: Node = some_html_element.into();
+x.append_child(some_node);
+```
+
+The end goal is that it should be possible to write generic code which can
+work with all sub-classes in the inheritance graph.
+
+As an example, it should be possible to use the
+[`appendChild`](https://developer.mozilla.org/en-US/docs/Web/API/Node/appendChild)
+method with all of the classes which inherit from `Node` (e.g. `HTMLElement`,
+`HTMLDivElement`, `SVGElement`, and many more). This should be possible without any
+casting operations (such as `into` or `dyn_into`):
+
+```rust
+some_html_element.append_child(some_node);
+```
+
+As a side effect, this also makes it possible to write generic code like this:
+
+```rust
+fn foo<A: INode>(node: A) { ... }
+```
+
+However, due to the potential for code bloat, this sort of pattern is *not* encouraged
+by this RFC.
+
+# Stakeholders
+[stakeholders]: #stakeholders
+
+Anyone who is using `wasm-bindgen` directly or transitively through the
+`web-sys` crate is affected. This does *not* affect the larger wasm ecosystem
+outside of Rust (eg Webpack). Therefore, the usual advertisement of this RFC on
+*This Week in Rust and WebAssembly* and at our working group meetings should
+suffice in soliciting feedback.
+
+# Detailed Explanation
+[detailed-explanation]: #detailed-explanation
+
+## Example Usage
+
+Let's consider adding in methods for the [`EventTarget`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget)
+and [`Node`](https://developer.mozilla.org/en-US/docs/Web/API/Node) classes.
+
+In WebIDL, they are specified like this:
+
+```
+[Constructor,
+ Exposed=(Window,Worker,AudioWorklet)]
+interface EventTarget {
+  void addEventListener(DOMString type, EventListener? callback, optional (AddEventListenerOptions or boolean) options);
+  void removeEventListener(DOMString type, EventListener? callback, optional (EventListenerOptions or boolean) options);
+  boolean dispatchEvent(Event event);
+};
+```
+
+```
+[Exposed=Window]
+interface Node : EventTarget {
+  [CEReactions] attribute DOMString? nodeValue;
+  [CEReactions] attribute DOMString? textContent;
+
+  [CEReactions] Node appendChild(Node node);
+  [CEReactions] Node removeChild(Node child);
+
+  ... other attributes and methods ommitted ...
+};
+```
+
+Based upon that WebIDL spec, the WebIDL generator will generate this Rust code:
```

Comment:
Small nitpick, the WebIDL frontend doesn't actual use the wasm-bindgen macros, it directly generates code. We often think about what is equivalent between the two, but they fundamentally both go directly to codegen. E.g. we would say that the snippet above might de-sugar into the extern block below, but it doesn't actually ever create the extern block below, it creates the same post-processed output of the extern block.